README.md
[![Build Status](https://travis-ci.org/cxw42/Sub-Multi-Tiny.svg?branch=master)](https://travis-ci.org/cxw42/Sub-Multi-Tiny)
# NAME

Sub::Multi::Tiny - Multisubs/multimethods (multiple dispatch) yet another way!

# SYNOPSIS

    {
        package main::my_multi;     # We're making main::my_multi()
        use Sub::Multi::Tiny qw($foo $bar);     # All possible params

        sub first :M($foo, $bar) {  # sub's name will be ignored,
            return $foo ** $bar;    # but can't match the one we're making
        }

        sub second :M($foo) {
            return $foo + 42;
        }

    }

    # Back in package main, my_multi() is created just before the run phase.
    say my_multi(2, 5);     # -> 32
    say my_multi(1295);     # -> 1337

**Limitation:** At present, dispatch is solely by arity, and only one
candidate can have each arity.  This limitation will be removed in the future.

# DESCRIPTION

Sub::Multi::Tiny is a library for making multisubs, aka multimethods,
aka multiple-dispatch subroutines.  Each multisub is defined in a
single package.  Within that package, the individual implementations ("impls")
are `sub`s tagged with the `:M` attribute.  The names of the impls are
preserved but not used specifically by Sub::Multi::Tiny.

Within a multisub package, the name of the sub being defined is available
for recursion.  For example (using `where`, not yet implemented):

    {
        package main::fib;
        use Sub::Multi::Tiny qw($n);
        sub base  :M($n where { $_ eq 0 })  { 1 }
        sub other :M($n)                    { $n * fib($n-1) }
    }

This code creates function `fib()` in package `main`.  Within package
`main::fib`, function `fib()` is an alias for `main::fib()`.  It's easier
to use than to explain!

# FUNCTIONS

## import

Sets up the package that uses it to define a multisub.  The parameters
are all the parameter variables that the multisubs will use.  `import`
creates these as package variables so that they can be used unqualified
in the multisub implementations.

A parameter `D:Dispatcher` can also be given to specify the dispatcher to
use.  If `Dispatcher` includes a double-colon, it will be used as a full
package name.  Otherwise, `Sub::Multi::Tiny::Dispatcher::` will be prepended.

Also sets ["$VERBOSE" in Sub::Multi::Tiny::Util](https://metacpan.org/pod/Sub::Multi::Tiny::Util#VERBOSE) if the environment variable
`SUB_MULTI_TINY_VERBOSE` has a truthy value.  If the `SUB_MULTI_TINY_VERBOSE`
value is numeric, `$VERBOSE` is set to that value; otherwise, `$VERBOSE` is
set to 1.

# CUSTOM DISPATCH

This module includes a default dispatcher (implemented in
[Sub::Multi::Tiny::DefaultDispatcher](https://metacpan.org/pod/Sub::Multi::Tiny::DefaultDispatcher).  To use a different dispatcher,
define or import a sub `MakeDispatcher()` into the package before
compilation ends.  That sub will be called to create the dispatcher.
For example:

    {
        package main::foo;
        use Sub::Multi::Tiny;
        sub MakeDispatcher { return sub { ... } }
    }

or

    {
        package main::foo;
        use Sub::Multi::Tiny;
        use APackageThatImportsMakeDispatcherIntoMainFoo;
    }

# DEBUGGING

For extra debug output, set ["$VERBOSE" in Sub::Multi::Tiny::Util](https://metacpan.org/pod/Sub::Multi::Tiny::Util#VERBOSE) to a positive
integer.  This has to be set at compile time to have any effect.  For example,
before creating any multisubs, do:

    use Sub::Multi::Tiny::Util '*VERBOSE';
    BEGIN { $VERBOSE = 2; }

# RATIONALE

- To be able to use multisubs in pre-5.14 Perls with only built-in
language facilities.  This will help me make my own modules backward
compatible with those Perls.
- To learn how it's done! :)

# SEE ALSO

I looked at these but decided not to use them for the following reasons:

- [Class::Multimethods](https://metacpan.org/pod/Class::Multimethods)

    I wanted a syntax that used normal `sub` definitions as much as possible.
    Also, I was a bit concerned by LPALMER's experience that it "does what you
    don't want sometimes without saying a word"
    (["Semantics" in Class::Multimethods::Pure](https://metacpan.org/pod/Class::Multimethods::Pure#Semantics)).

    Other than that, I think this looks pretty decent (but haven't tried it).

- [Class::Multimethods::Pure](https://metacpan.org/pod/Class::Multimethods::Pure)

    Same desire for `sub` syntax.  Additionally, the last update was in 2007,
    and the maintainer hasn't uploaded anything since.  Other than that, I think
    this also looks like a decent option (but haven't tried it).

- [Dios](https://metacpan.org/pod/Dios)

    This is a full object system, which I do not need in my use case.

- [Logic](https://metacpan.org/pod/Logic)

    This one is fairly clean, but uses a source filter.  I have not had much
    experience with source filters, so am reluctant.

- [Kavorka::Manual::MultiSubs](https://metacpan.org/pod/Kavorka::Manual::MultiSubs) (and [Moops](https://metacpan.org/pod/Moops))

    Requires Perl 5.14+.

- [MooseX::MultiMethods](https://metacpan.org/pod/MooseX::MultiMethods)

    I am not ready to move to full [Moose](https://metacpan.org/pod/Moose)!

- [MooseX::Params](https://metacpan.org/pod/MooseX::Params)

    As above.

- [Sub::Multi](https://metacpan.org/pod/Sub::Multi)

    The original inspiration for this module, whence this module's name.
    `Sub::Multi` uses coderefs, and I wanted a syntax that used normal
    `sub` definitions as much as possible.

- [Sub::SmartMatch](https://metacpan.org/pod/Sub::SmartMatch)

    This one looks very interesting, but I haven't used smartmatch enough
    to be fully comfortable with it.

# SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Sub::Multi::Tiny

You can also look for information at:

- GitHub: The project's main repository and issue tracker

    [https://github.com/cxw42/Sub-Multi-Tiny](https://github.com/cxw42/Sub-Multi-Tiny)

- MetaCPAN

    [Sub::Multi::Tiny](https://metacpan.org/pod/Sub::Multi::Tiny)

- This distribution

    See the tests in the `t/` directory distributed with this software
    for examples.

# BUGS

This isn't Damian code ;) .

# AUTHOR

Chris White <cxw@cpan.org>

# LICENSE

Copyright (C) 2019 Chris White <cxw@cpan.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Perl extension Sub-Multi-Tiny

{{$NEXT}}

0.000013 2022-10-30T01:15:39Z

    - Update per upstream Type::Params changes (#10; thanks to TOBYINK)

0.000012 2019-09-24T01:52:46Z

    - Internal improvements for robustness.
    - S::M::T::Util::_complete_dispatcher now adds strict and warnings
      to the top of the generated dispatcher.

0.000011 2019-09-20T21:41:16Z

    - Fix POD typo (thanks to MANWAR)

0.000010 2019-09-20T02:36:46Z (TRIAL RELEASE)

    - Add tests
    - Refactor out common test code
    - In D:Default, die loudly on type or where checks, rather
      than ignoring them silently.
    - Update documentation

0.000009 2019-09-14T02:28:08Z (TRIAL RELEASE)

    - Fix Dispatcher::TypeParams to run on older Perls (#8)

0.000008 2019-09-13T22:16:40Z (TRIAL RELEASE)

    - Fix t/Kit.pm to run on older Perls

0.000007 2019-09-12T01:13:30Z (TRIAL RELEASE)

    - The SUB_MULTI_TINY_VERBOSE environment value can have a numeric value
    - New dependency on Scalar::Util
    - Added voluminous test output on earlier Perls in hopes of tracking
        down cpantesters failures.
    - Use t/Kit.pm consistently in test files
    - Verbose test output while covering, so the lines that provide that
        output are covered!

0.000006 2019-09-11T00:46:14Z

    - Added a Type::Params-based dispatcher
    - Added pluggable dispatcher loader
    - Fixed cpantesters failures on mswin32 and 5.8.6/5.8.7

0.000005 2019-09-06T02:03:22Z

    - Changes to cpanfile, xt to try to fix cpantesters failures (#5)

0.000004 2019-09-05T02:14:14Z

    - Support recursion to the dispatcher (#4)
    - Add documentation
    - Relax the minimum Perl version to 5.006

0.000003 2019-09-04T14:21:27Z

    - Updated documentation
    - Added tests
    - Removed undeclared dependency (#1)

0.000002 2019-09-03T01:00:33Z

    - Actually pass parameter values to the correct candidate!

0.000001 2019-09-01T01:26:34Z

    - Initial release to CPAN.  Only supports dispatch by arity.

# vi: set ts=4 sts=4 sw=4 et ai: #


t/Kit.pm
# Kit.pm: test kit for Sub::Multi::Tiny
package # hide from PAUSE
    Kit;

use 5.006;
use strict;
use warnings;

use parent 'Exporter';
use vars::i '@EXPORT' => qw(find_file_in_t get_perl_filename is_covering
                            run_perl true false);

use Config;
use Cwd 'abs_path';
use Data::Dumper;
use File::Spec;
use Import::Into;
use IPC::Run3;
use Test::More;

# is_covering: is Devel::Cover running?
sub is_covering {
    return !!(eval 'Devel::Cover::get_coverage()');
} #is_covering()

# Set verbosity:
#   - on for coverage of _hlog statements
#   - on for debugging of test failures in <5.18    XXX DEBUG
# Note that verbosity will also be on if $ENV{SUB_MULTI_TINY_VERBOSE} is set.
use Sub::Multi::Tiny::Util '*VERBOSE';
BEGIN {
    $VERBOSE = 99 if is_covering || $] lt '5.018';
}

use constant {
    true => !!1,
    false => !!0,
};


# Get the filename of the Perl interpreter running this. {{{1
# Modified from perlvar.
# The -x test is for cygwin or other systems where $Config{perlpath} has no
# extension and $Config{_exe} is nonempty.  E.g., symlink perl->perl5.10.1.exe.
# There is no "perl.exe" on such a system.
sub get_perl_filename {
    my $secure_perl_path = $Config{perlpath};
    if ($^O ne 'VMS') {
        $secure_perl_path .= $Config{_exe}
            unless (-x $secure_perl_path) ||
                            ($secure_perl_path =~ m/$Config{_exe}$/i);
    }
    die "Could not find perl interpreter" unless $secure_perl_path;
    return $secure_perl_path;
} # get_perl_filename()

# }}}1

# find_file_in_t($filename[, checks]).  Assumes caller is in t/.
sub find_file_in_t {
    my (undef, $filename) = caller;

    my $here = abs_path($filename);
    die "Could not find my file location: $!" unless defined $here;
    my ($volume,$directories,undef) = File::Spec->splitpath( $here );

    my $pl_file = File::Spec->catpath(
        $volume,
        $directories,
        shift
    );

    # File tests, if requested.
    foreach(@_) {
        die "Can not read $pl_file" if $_ eq 'r' && !( -f $pl_file && -r _);
        die "Can not write $pl_file" if $_ eq 'w' && !( -w $pl_file );
        die "Can not execute $pl_file" if $_ eq 'x' && !( -f $pl_file && -x _);
    }

    return $pl_file;
} #find_file_in_t

# ($out, $err, $exitstatus) = run_perl(args arrayref, [$stdin text if any])
sub run_perl {
    my $perl = get_perl_filename;
    my ($lrArgs, $in) = @_;
    $in = '' unless defined $in;
    my ($out, $err);

    # Check if we are running under cover(1) from Devel::Cover
    diag is_covering() ? 'Devel::Cover running' : 'Devel::Cover not covering';

    # Note: See App-PRT/t/App-PRT-CLI.t for code to find test scripts in
    # script vs. blib/script, if that later becomes necessary.

    # Make the command to run script/prt.
    my @cmd = ($perl, is_covering() ? ('-MDevel::Cover=-silent,1') : ());

    push @cmd, map { "-I$_" } @INC;
    push @cmd, @$lrArgs;
    diag 'Running ', join ' ', @cmd;
    run3 \@cmd, \$in, \$out, \$err;     # Dies on error

    my $exitstatus = $?;
    diag "Error message was '$err'" if $err;
    return ($out, $err, $exitstatus);
} #run_perl

# Import
sub import {
    my ($target, $filename) = caller;
    __PACKAGE__->export_to_level(1, @_);
    $_->import::into($target) foreach qw(strict warnings Config Cwd
        File::Spec IPC::Run3 Test::More);

    $Data::Dumper::Indent = 1;  # fixed indentation per level

    diag '#' x 40;
    diag $filename;
}

1;
# Documentation {{{1
__END__

=head1 NAME

Kit - Test kit for Sub::Multi::Tiny

=head1 AUTHOR

Chris White E<lt>cxw@cpan.orgE<gt>

=head1 LICENSE

Copyright (C) 2019 Chris White E<lt>cxw@cpan.orgE<gt>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

# }}}1
# vi: set fdm=marker: #


lib/Sub/Multi/Tiny/SigParse.pm
####################################################################
#
#    This file was generated using Parse::Yapp version 1.21.
#
#        Don't edit this file, use source file instead.
#
#             ANY CHANGE MADE HERE WILL BE LOST !
#
####################################################################
package Sub::Multi::Tiny::SigParse;
use vars qw ( @ISA );
use strict;

@ISA= qw ( Parse::Yapp::Driver );
use Parse::Yapp::Driver;

#line 6 "/cygdrive/c/Users/cxw/proj/Sub-Multi-Tiny/support/SigParse.yp"


# Imports {{{1

use 5.010001;
use strict;
use warnings;

# }}}1
# Documentation {{{1

=head1 NAME

Sub::Multi::Tiny::SigParse - Parse::Yapp input to parse signatures in Sub::Multi::Tiny

=head1 SYNOPSIS

Generate the .pm file:

    yapp -m Sub::Multi::Tiny::SigParse -o lib/Sub/Multi/Tiny/SigParse.pm support/SigParse.yp

And then:

    use Sub::Multi::Tiny::SigParse;
    my $ast = Sub::Multi::Tiny::SigParse::Parse($signature);

=head1 FUNCTIONS

=cut

# }}}1



sub new {
        my($class)=shift;
        ref($class)
    and $class=ref($class);

    my($self)=$class->SUPER::new( yyversion => '1.21',
                                  yystates =>
[
	{#State 0
		DEFAULT => -1,
		GOTOS => {
			'signature' => 1
		}
	},
	{#State 1
		ACTIONS => {
			'' => 2
		}
	},
	{#State 2
		DEFAULT => 0
	}
],
                                  yyrules  =>
[
	[#Rule 0
		 '$start', 2, undef
	],
	[#Rule 1
		 'signature', 0,
sub
#line 54 "/cygdrive/c/Users/cxw/proj/Sub-Multi-Tiny/support/SigParse.yp"
{ +{} }
	]
],
                                  @_);
    bless($self,$class);
}

#line 58 "/cygdrive/c/Users/cxw/proj/Sub-Multi-Tiny/support/SigParse.yp"


#############################################################################
# Footer

# Tokenizer and error-reporting routine for Parse::Yapp {{{1

# The lexer
sub _next_token {
    my $parser = shift;
    my $text = $parser->YYData->{TEXT};
    return ('', undef) unless $text;    # EOF

    ...
} #_next_token()

# Report an error
sub _report_error {
    my $parser = shift;
    my $got = $parser->YYCurtok || '<end of input>';
    my $val='';
    $val = ' (' . $parser->YYCurval . ')' if $parser->YYCurval;
    die 'Syntax error: could not understand ', $got, $val, "\n";
    if(ref($parser->YYExpect) eq 'ARRAY') {
        print 'Expected one of: ', join(',', @{$parser->YYExpect}), "\n";
    }
    return;
} #_report_error()

# }}}1
# Top-level parse function {{{1

=head2 Parse

Parse arguments.  Usage:

    my $ast = Sub::Multi::Tiny::SigParse::Parse($signature);

=cut

sub Parse {
    my $text = shift or
        (require Carp, Carp::croak 'Parse: Need a signature to parse');

    my $parser = __PACKAGE__->new;
    my $hrData = $parser->YYData;

    # Data we use while parsing
    $hrData->{TEXT} = $text;

    my $hrRetval = $parser->YYParse(yylex => \&_next_token,
        yyerror => \&_report_error,
        (@_ ? (yydebug => $_[0]) : ()),
    );

    return $hrRetval;
} #Parse()

# }}}1
# Rest of the docs {{{1

=head1 AUTHOR

Chris White E<lt>cxw@cpan.orgE<gt>

=head1 LICENSE

Copyright (C) 2019 Chris White E<lt>cxw@cpan.orgE<gt>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

# }}}1

# vi: set fdm=marker: #

1;


lib/Sub/Multi/Tiny.pm
package Sub::Multi::Tiny;

use 5.006;
use strict;
use warnings;

require Attribute::Handlers;    # Listed here so automated tools see it

use Import::Into;
use Scalar::Util qw(looks_like_number);
use Sub::Multi::Tiny::SigParse;
use Sub::Multi::Tiny::Util ':all';
use subs ();
use vars ();

our $VERSION = '0.000010'; # TRIAL

use constant { true => !!1, false => !!0 };

# Documentation {{{1

=encoding utf-8

=head1 NAME

Sub::Multi::Tiny - Multisubs/multimethods (multiple dispatch) yet another way!

=head1 SYNOPSIS

    {
        package main::my_multi;     # We're making main::my_multi()
        use Sub::Multi::Tiny qw($foo $bar);     # All possible params

        sub first :M($foo, $bar) {  # sub's name will be ignored,
            return $foo ** $bar;    # but can't match the one we're making
        }

        sub second :M($foo) {
            return $foo + 42;
        }

    }

    # Back in package main, my_multi() is created just before the run phase.
    say my_multi(2, 5);     # -> 32
    say my_multi(1295);     # -> 1337

The default dispatcher dispatches solely by arity, and only one
candidate can have each arity.  For more flexible dispatching, see
L<Sub::Multi::Tiny::Dispatcher::TypeParams>.

=head1 DESCRIPTION

Sub::Multi::Tiny is a library for making multisubs, aka multimethods,
aka multiple-dispatch subroutines.  Each multisub is defined in a
single package.  Within that package, the individual implementations ("impls")
are C<sub>s tagged with the C<:M> attribute.  The names of the impls are
preserved but not used specifically by Sub::Multi::Tiny.

Within a multisub package, the name of the sub being defined is available
for recursion.  For example (using C<where>, supported by
L<Sub::Multi::Tiny::Dispatcher::TypeParams>):

    {
        package main::fib;
        use Sub::Multi::Tiny qw(D:TypeParams $n);
        sub base  :M($n where { $_ <= 1 })  { 1 }
        sub other :M($n)                    { $n * fib($n-1) }
    }

This code creates function C<fib()> in package C<main>.  Within package
C<main::fib>, function C<fib()> is an alias for C<main::fib()>.  It's easier
to use than to explain!

=head1 FUNCTIONS

=cut

# }}}1

# Information about the multisubs so we can create the dispatchers at
# INIT time.
my %_multisubs;

# Sanity check: any :M will die after the INIT block below runs.
my $_dispatchers_created;

# Accessor
sub _dispatchers_created { !!$_dispatchers_created; }

# INIT: Fill in the dispatchers for any multisubs we've created.
# Note: attributes are applied at CHECK time, before this.
# We use INIT so that compilation failures will prevent this code from running.

INIT {
    _hlog { __PACKAGE__, "in INIT block" };
    $_dispatchers_created = 1;
    while(my ($multisub_fullname, $hr) = each(%_multisubs)) {
        my $dispatcher = _make_dispatcher($hr)
            or die "Could not create dispatcher for $multisub_fullname\()";

        eval { no strict 'refs'; *{$multisub_fullname} = $dispatcher };
        die "Could not assign dispatcher for $multisub_fullname\:\n$@" if $@;
        do {
            no strict 'refs';
            no warnings 'redefine';
            my $target_name = "$hr->{defined_in}\::$hr->{subname}";
            *{$target_name} = $dispatcher;
        };
    } #foreach multisub
} #CHECK

=head2 import

Sets up the package that uses it to define a multisub.  The parameters
are all the parameter variables that the multisubs will use.  C<import>
creates these as package variables so that they can be used unqualified
in the multisub implementations.

A parameter C<D:Dispatcher> can also be given to specify the dispatcher to
use --- see L</CUSTOM DISPATCH>.

Also sets L<Sub::Multi::Tiny::Util/$VERBOSE> if the environment variable
C<SUB_MULTI_TINY_VERBOSE> has a truthy value.  If the C<SUB_MULTI_TINY_VERBOSE>
value is numeric, C<$VERBOSE> is set to that value; otherwise, C<$VERBOSE> is
set to 1.

=cut

sub import {
    my $multi_package = caller;     # The package that defines the multisub
    my $my_package = shift;         # The package we are

    for($ENV{SUB_MULTI_TINY_VERBOSE}) {
        last unless $_;
        $VERBOSE = looks_like_number($_) ? 0+ $_ : 1;
    }

    if(@_ && $_[0] eq ':nop') {
        _hlog { __PACKAGE__ . ':nop => Taking no action' } 0;    # Always
        return;
    }

    _hlog { "Target $multi_package package $my_package" };
    my ($target_package, $subname) = ($multi_package =~ m{^(.+?)::([^:]+)$});
        # $target_package is the package that will be able to call the multisub
    _croak "Can't parse package name ${multi_package} into <target>::<name>"
        unless $target_package && $subname;

    _croak "Can't redefine multi sub $multi_package\()"
        if exists $_multisubs{$multi_package};

    # Create the vars - they will be accessed as package variables.
    # TODO: parameters of the form D:<foo> import dispatcher <foo>.
    my @possible_params;
    my $dispatcher = 'Default';
    foreach (@_) {
        if(/^D:(.*)$/) {
            die '"D:" must be followed by a dispatcher class' unless $1;
            $dispatcher=$1;
        } elsif(/^.:/) {
            die '".:..." forms reserved - did you mean "D:DispatcherClass"?';
        } else {
            push @possible_params, $_;
        }
    }

    # Load the parameter variables as package variables
    _croak "Please list the sub parameters" unless @possible_params;
    vars->import::into($multi_package, @possible_params);

    # Load the dispatcher
    $dispatcher = __PACKAGE__ . "::Dispatcher::$dispatcher"
        unless index($dispatcher, '::') != -1;

    eval "require $dispatcher";
    die "Could not load dispatcher $dispatcher, requested by $multi_package: $@"
        if $@;

    _hlog { $multi_package, 'using dispatcher', $dispatcher };
    ${dispatcher}->import::into($multi_package);

    # Make a stub that we will redefine later
    _hlog { "Making $multi_package\()" } ;
    subs->import::into($target_package, $subname);
    # TODO add stub for callsame/nextwith/...

    # Save the patch
    $_multisubs{$multi_package} = {
        used_by => $target_package,
        defined_in => $multi_package,
        subname => $subname,
        possible_params => +{ map { ($_ => 1) } @possible_params },
        impls => [],    # Implementations - subs tagged :M
    };

    # Set up the :M attribute in $multi_package if it doesn't
    # exist yet.
    if(eval { no strict 'refs'; defined &{$multi_package . '::M'} }) {
        die "Cannot redefine M in $multi_package";
    } else {
        _hlog { "Making $multi_package attr M" } 2;
        eval(_make_M($multi_package));
        die $@ if $@;
    }

    # Set up $subname() in $multi_package, which will be aliased to the
    # dispatcher.
    if(eval { no strict 'refs'; defined &{"$multi_package\::$subname"} }) {
        die "Cannot redefine $subname in $multi_package";
    } else {
        _hlog { "Making $multi_package\::$subname stub" } 2;
        do { no strict 'refs'; *{"$multi_package\::$subname"} = sub {} };
    }
} #import()

# Parse the argument list to the attribute handler
sub _parse_arglist {
    my ($spec, $funcname) = @_;
    _croak "Need a parameter spec for $funcname" unless $spec;
    _hlog { "Parsing args for $funcname: $spec" } 2;

    return Sub::Multi::Tiny::SigParse::Parse($spec);
} #_parse_arglist

# Create the source for the M attribute handler for a given package
sub _make_M {
    my $multi_package = shift;
    my $P = __PACKAGE__;
    my $code = _line_mark_string
        "package $multi_package;\n";

    # TODO See if making M an :ATTR(..., BEGIN) permits us to remove the
    # requirement to list all the parameters in the `use S::M::T` line

    $code .= _line_mark_string <<'EOT';
use Attribute::Handlers;
use Sub::Multi::Tiny::Util qw(_hlog);
##use Data::Dumper;

sub M :ATTR(CODE,RAWDATA) {
    _hlog { require Data::Dumper;
            'In ', __PACKAGE__, "::M: \n",
            Data::Dumper->Dump([\@_], ['attr_args']) } 2;

    my ($package, $symbol, $referent, $attr, $data, $phase,
        $filename, $linenum) = @_;
    my $funcname = "$package\::" . *{$symbol}{NAME};

    _hlog {     # Code from Attribute::Handlers, license perl_5
        ref($referent),
        $funcname,
        "($referent)", "was just declared",
        "and ascribed the ${attr} attribute",
        "with data ($data)",
        "in phase $phase",
        "in file $filename at line $linenum"
    } 2;
EOT

    # Trap out-of-sequence calls.  Currently you can't create a new multisub
    # via eval at runtime.  TODO use UNITCHECK instead to permit doing so?
    $code .= _line_mark_string <<EOT;
    die 'Dispatchers already created - please file a bug report'
        if $P\::_dispatchers_created();

    my \$multi_def = \$_multisubs{'$multi_package'};
EOT

    # Parse and validate the args
    $code .= _line_mark_string <<EOT;
    my \$hrSig = $P\::_parse_arglist(\$data, \$funcname);
    $P\::_check_and_inflate_sig(\$hrSig, \$multi_def,
        \$funcname, \$package, \$filename, \$linenum);
EOT

    $code .= _line_mark_string <<'EOT';
EOT

    # Save the implementation's info for use when making the dispatcher.
    $code .= _line_mark_string <<'EOT';
    my $info = {
        code => $referent,
        args => $hrSig->{parms},    # TODO remove eventually
        sig => $hrSig,

        # For error messages
        filename => $filename,
        linenum => $linenum,
        candidate_name => $funcname
    };
    push @{$multi_def->{impls}}, $info;

} #M
EOT

    _hlog { "M code:\n$code\n" } 2;
    return $code;
} #_make_M

# Validate a signature and convert text to usable objects
sub _check_and_inflate_sig {
    my ($signature, $multi_def, $funcname, $package, $filename, $linenum) = @_;
    my ($saw_positional, $saw_named);

    my $args = $signature->{parms};
    my $temp;
    foreach (@$args) {

        # Is the argument valid in this package?
        my $name = $_->{name};
        unless($multi_def->{possible_params}->{$name}) {
            die "Argument $name is not listed on the 'use Sub::Multi::Tiny' line (used by $funcname at $filename\:$linenum";
        }

        # Is the argument out of order?
        die "Positional arguments must precede named arguments"
            if $saw_named && !$_->{named};

        # Inflate type constraint, if any
        if($_->{type}) {
            _hlog { In => $package, "evaluating type '$_->{type}'" };
            $temp = eval _line_mark_string <<EOT;
do {
    package $package;
    $_->{type}  # Anything meaningful in the calling package is OK
}
EOT
            die "In $package: Could not understand type '$_->{type}': $@" if $@;
            $_->{type} = $temp;
        }

        # Inflate where clause, if any, into a closure
        if($_->{where}) {
            _hlog { In => $package, "evaluating 'where' clause '$_->{where}'" };
            $temp = eval _line_mark_string <<EOT;
do {
    package $package;
    sub $_->{where}     # Anything meaningful in the calling package is OK
}
EOT
            die "In $package: Could not understand 'where' clause '$_->{where}': $@" if $@;
            $_->{where} = $temp;
        }

        # Remember data for later
        $saw_named ||= $_->{named};
        $saw_positional ||= !$_->{named};

    }
} # _check_and_inflate_sig

# Create a dispatcher
sub _make_dispatcher {
    my $hr = shift;
    die "No implementations given for $hr->{defined_in}"
        unless @{$hr->{impls}};

    my $custom_dispatcher = do {
        no strict 'refs';
        *{ $hr->{defined_in} . '::MakeDispatcher' }{CODE}
    };

    return $custom_dispatcher->($hr) if defined $custom_dispatcher;

    # Default dispatcher
    require Sub::Multi::Tiny::Dispatcher::Default;
    return Sub::Multi::Tiny::Dispatcher::Default::MakeDispatcher($hr);
} #_make_dispatcher

1;
# Rest of the documentation {{{1
__END__

=head1 CUSTOM DISPATCH

This module includes a default dispatcher (implemented in
L<Sub::Multi::Tiny::Dispatcher::Default>.  To use a different dispatcher,
define or import a sub C<MakeDispatcher()> into the package before
compilation ends.  That sub will be called to create the dispatcher.
For example:

    {
        package main::foo;
        use Sub::Multi::Tiny;
        sub MakeDispatcher { return sub { ... } }
    }

or

    {
        package main::foo;
        use Sub::Multi::Tiny;
        use APackageThatImportsMakeDispatcherIntoMainFoo;
    }

As a shortcut, you can specify a dispatcher on the C<use> line.  For example:

    use Sub::Multi::Tiny qw(D:Foo $var);

will use dispatcher C<Sub::Multi::Tiny::Dispatcher::Foo>.  Any name with a
double-colon will be used as a full package name.  E.g., C<D:Bar::Quux> will
use dispatcher C<Bar::Quux>.  If C<Foo> does not include a double-colon,
C<Sub::Multi::Tiny::Dispatcher::> will be prepended.

=head1 DEBUGGING

For extra debug output, set L<Sub::Multi::Tiny::Util/$VERBOSE> to a positive
integer.  This has to be set at compile time to have any effect.  For example,
before creating any multisubs, do:

    use Sub::Multi::Tiny::Util '*VERBOSE';
    BEGIN { $VERBOSE = 2; }

=head1 RATIONALE

=over

=item *

To be able to use multisubs in pre-5.14 Perls with only built-in
language facilities.  This will help me make my own modules backward
compatible with those Perls.

=item *

To learn how it's done! :)

=back

=head1 SEE ALSO

I looked at these but decided not to use them for the following reasons:

=over

=item L<Class::Multimethods>

I wanted a syntax that used normal C<sub> definitions as much as possible.
Also, I was a bit concerned by LPALMER's experience that it "does what you
don't want sometimes without saying a word"
(L<Class::Multimethods::Pure/Semantics>).

Other than that, I think this looks pretty decent (but haven't tried it).

=item L<Class::Multimethods::Pure>

Same desire for C<sub> syntax.  Additionally, the last update was in 2007,
and the maintainer hasn't uploaded anything since.  Other than that, I think
this also looks like a decent option (but haven't tried it).

=item L<Dios>

This is a full object system, which I do not need in my use case.

=item L<Logic>

This one is fairly clean, but uses a source filter.  I have not had much
experience with source filters, so am reluctant.

=item L<Kavorka::Manual::MultiSubs> (and L<Moops>)

Requires Perl 5.14+.

=item L<MooseX::MultiMethods>

I am not ready to move to full L<Moose>!

=item L<MooseX::Params>

As above.

=item L<Sub::Multi>

The original inspiration for this module, whence this module's name.
C<Sub::Multi> uses coderefs, and I wanted a syntax that used normal
C<sub> definitions as much as possible.

=item L<Sub::SmartMatch>

This one looks very interesting, but I haven't used smartmatch enough
to be fully comfortable with it.

=back

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Sub::Multi::Tiny

You can also look for information at:

=over

=item * GitHub: The project's main repository and issue tracker

L<https://github.com/cxw42/Sub-Multi-Tiny>

=item * MetaCPAN

L<Sub::Multi::Tiny>

=item * This distribution

See the tests in the C<t/> directory distributed with this software
for usage examples.

=back

=head1 BUGS

=over

=item * It's not as tiny as I thought it would be!

=item * This isn't Damian code ;) .

=back

=head1 AUTHOR

Chris White E<lt>cxw@cpan.orgE<gt>

=head1 LICENSE

Copyright (C) 2019 Chris White E<lt>cxw@cpan.orgE<gt>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

# }}}1
# vi: set fdm=marker: #
